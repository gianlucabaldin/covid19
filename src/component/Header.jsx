import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import GitHubIcon from '@material-ui/icons/GitHub';
import { Box } from '@material-ui/core';
import NavButtons from './NavButtons';

const useStyles = makeStyles(() => ({
  toolbar: {
    justifyContent: 'space-between',
  },
}));

const Header = () => {
  const classes = useStyles();

  return (
    <AppBar position="static" data-id="appbar" color="transparent">
      <Toolbar className={classes.toolbar}>
        <Box width={200}>
          <Typography variant="h4" color="primary">
            Covid 19
          </Typography>
          <Typography variant="p" color="primary">
            Latest & historical data
          </Typography>
        </Box>
        <Box>
          <NavButtons />
        </Box>
        <Box width={200} textAlign="right">
          <GitHubIcon
            href="https://github.com/gianlucabaldin"
            fontSize="large"
          />
        </Box>
      </Toolbar>
    </AppBar>
  );
};

export default Header;
