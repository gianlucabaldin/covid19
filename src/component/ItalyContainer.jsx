/* eslint-disable camelcase */
/* eslint-disable react/jsx-props-no-spreading */
import React, { useEffect, useState } from 'react';
import { Grid, Typography } from '@material-ui/core';
import moment from 'moment';
import { useTranslation } from 'react-i18next';
import ChartContainer from './ChartContainer';
import ItalyRegions from './ItalyRegions';
import { fetchItalyRegion, fetchItalyHistoricalAll } from '../utils/fetch';
import Summary from './Summary';
import { processData } from '../utils/chartUtils';
import { SECTIONS, API_APIFY_SHORT_URL } from '../utils/consts';

const summaryInitialStatus = {
  tamponi: 0,
  totalCases: 0,
  deceased: 0,
  lastUpdate: '',
  dataTestId: SECTIONS.ITALY,
  loading: true,
  error: false,
};

const chartInitialStatus = {
  data: [],
  dataTestId: SECTIONS.ITALY,
  error: false,
  loading: true,
  checked: true,
  href: API_APIFY_SHORT_URL,
};

const tableInitialStatud = {
  data: {},
  loading: true,
};

const ItalyContainer = () => {
  const { t } = useTranslation();

  const [summaryData, setSummaryData] = useState({ ...summaryInitialStatus });
  const [chartData, setChartData] = useState({ ...chartInitialStatus });
  const [tableData, setTableData] = useState(tableInitialStatud);
  const [fetchedDataAll, setFetchedDataAll] = useState();

  const extractData = (res) => {
    const {
      intensiveTherapy,
      totalHospitalized,
      totalPositive,
      newDailyPositive,
      dailyDeceased,
      dailySwabs,
    } = processData(res);

    // fill summary with last day-data extracted from previous fetch
    setSummaryData({
      ...summaryData,
      data: [
        { key: 'total-swabs', value: res[res.length - 1].tamponi },
        { key: 'total-cases', value: res[res.length - 1].totalCases },
        { key: 'total-deceased', value: res[res.length - 1].deceased },
        {
          key: 'last-update',
          value: moment(intensiveTherapy[intensiveTherapy.length - 1].x).format(
            'LLL',
          ),
        },
      ],
      loading: false,
      error: false,
    });
    setChartData({
      ...chartInitialStatus,
      data: [
        {
          'intensive-therapy': intensiveTherapy,
        },
        { 'total-hospitalized': totalHospitalized },
        { 'total-positive': totalPositive },
        { 'new-daily-positive': newDailyPositive },
        { 'daily-deceased': dailyDeceased },
        { 'daily-swabs': dailySwabs },
      ],
      loading: false,
    });
    return res;
  };

  // fetch data from public api or mock (see implementation)
  const fetchData = () => {
    fetchItalyHistoricalAll()
      .then((res) => {
        return extractData(res);
      })
      .then((res) => {
        setFetchedDataAll(res);
      })
      .catch(() => {
        setSummaryData({ loading: false, error: true });
        setChartData({ error: true });
      });

    // fill table with fetched data
    fetchItalyRegion()
      .then((res) => {
        setTableData({ data: res, loading: false });
      })
      .catch((err) => {
        // to fill
        console.log('error', err);
      });
  };
  useEffect(() => {
    fetchData();
  }, []);

  const onToggleSwitch = (checked) => {
    const {
      intensiveTherapy,
      totalHospitalized,
      totalPositive,
      newDailyPositive,
      dailyDeceased,
      dailySwabs,
    } = processData(fetchedDataAll, checked);

    setChartData({
      ...chartInitialStatus,
      data: [
        {
          'intensive-therapy': intensiveTherapy,
        },
        { 'total-hospitalized': totalHospitalized },
        { 'total-positive': totalPositive },
        { 'new-daily-positive': newDailyPositive },
        { 'daily-deceased': dailyDeceased },
        { 'daily-swabs': dailySwabs },
      ],
      loading: false,
      checked,
    });
  };

  return (
    <div data-test-id="italy-container">
      <Grid container justify="center">
        <Summary {...summaryData} />
      </Grid>
      <Grid
        container
        style={{ paddingLeft: 16, paddingRight: 16 }}
        justify="center"
      >
        <Grid item>
          <ChartContainer {...chartData} onToggleSwitch={onToggleSwitch} />
        </Grid>
      </Grid>
      <Grid
        container
        style={{ paddingLeft: 16, paddingRight: 16 }}
        justify="center"
      >
        <Grid item style={{ marginTop: 10 }}>
          <Typography variant="h5" style={{ textTransform: 'capitalize' }}>
            {t('regions-details')}
          </Typography>
        </Grid>
      </Grid>
      <Grid
        container
        style={{ paddingLeft: 16, paddingRight: 16 }}
        justify="center"
      >
        <Grid item style={{ marginTop: 10 }}>
          <ItalyRegions {...tableData} />
        </Grid>
      </Grid>
    </div>
  );
};

// ItalyContainer.defaultProps = {
//   data: {},
//   error: false,
// };

export default ItalyContainer;
