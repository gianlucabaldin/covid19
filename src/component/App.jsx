/* eslint-disable import/no-cycle */
import React, { useState } from 'react';
import ItalyContainer from './ItalyContainer';
import Header from './Header';

export const SECTIONS = {
  WORLDWIDE: 'WORLDWIDE',
  ITALY: 'ITALY',
  COUNTRY_LIST: 'COUNTRY_LIST',
  EUROPE: 'EUROPE',
};

const App = () => {
  const [activeSection, setActiveSection] = useState(SECTIONS.ITALY);
  const handleClick = (location) => {
    if (location) setActiveSection(location);
  };
  return (
    <div>
      <Header onClick={handleClick} />
      {activeSection === SECTIONS.ITALY && <ItalyContainer />}
      {activeSection === SECTIONS.WORLDWIDE && <div>WORLDWIDE</div>}
      {activeSection === SECTIONS.COUNTRY_LIST && <div>COUNTRY_LIST</div>}
      {activeSection === SECTIONS.EUROPE && <div>EUROPE</div>}
    </div>
  );
};

export default App;
