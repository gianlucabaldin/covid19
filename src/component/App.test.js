import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import Header from './Header';
import ItalyContainer from './ItalyContainer';

describe('App', () => {
  const wrapper = shallow(<App />);

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('contains', () => {
    it('contains an Header component', () => {
      expect(wrapper.find(Header).length).toEqual(1);
    });

    it('contains ItalyContainer component', () => {
      expect(wrapper.find(ItalyContainer).length).toEqual(1);
      expect(wrapper.contains(<ItalyContainer />)).toEqual(true);
    });
  });
});
