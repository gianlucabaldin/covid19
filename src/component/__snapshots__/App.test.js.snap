// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div>
  <Header />
  <ItalyContainer
    data={Object {}}
    error={false}
  />
</div>
`;
